File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoMiddleware/Mistral.swift
import Foundation
import IAIsoCore

/// Mistral chat middleware.
public enum Mistral {

    public protocol Client {
        func chatComplete(_ params: [String: Any]) throws -> Response
    }

    public struct Usage: Sendable, Equatable {
        public let promptTokens: Int64
        public let completionTokens: Int64
        public let totalTokens: Int64
        public init(
            promptTokens: Int64 = 0,
            completionTokens: Int64 = 0,
            totalTokens: Int64 = 0
        ) {
            self.promptTokens = promptTokens
            self.completionTokens = completionTokens
            self.totalTokens = totalTokens
        }
    }

    public struct Choice: Sendable {
        public let toolCalls: [String]
        public init(toolCalls: [String] = []) { self.toolCalls = toolCalls }
    }

    public struct Response: Sendable {
        public let model: String
        public let usage: Usage
        public let choices: [Choice]
        public init(model: String, usage: Usage, choices: [Choice]) {
            self.model = model
            self.usage = usage
            self.choices = choices
        }
    }

    public struct Options: Sendable {
        public var raiseOnEscalation: Bool
        public init(raiseOnEscalation: Bool = false) {
            self.raiseOnEscalation = raiseOnEscalation
        }
        public static let defaults = Options()
    }

    public final class BoundedClient {
        private let raw: Client
        private let execution: BoundedExecution
        private let opts: Options

        public init(raw: Client, execution: BoundedExecution, options: Options = .defaults) {
            self.raw = raw
            self.execution = execution
            self.opts = options
        }

        public func chatComplete(_ params: [String: Any]) throws -> Response {
            let pre = execution.check()
            if pre == .locked { throw MiddlewareError.locked }
            if pre == .escalated, opts.raiseOnEscalation {
                throw MiddlewareError.escalationRaised
            }
            let resp: Response
            do {
                resp = try raw.chatComplete(params)
            } catch {
                throw MiddlewareError.provider(
                    error.localizedDescription, underlying: error)
            }
            var tokens = resp.usage.totalTokens
            if tokens == 0 {
                tokens = resp.usage.promptTokens + resp.usage.completionTokens
            }
            let toolCalls: Int64 = resp.choices.reduce(0) { acc, c in
                acc + Int64(c.toolCalls.count)
            }
            let model = resp.model.isEmpty ? "unknown" : resp.model
            execution.recordStep(StepInput(
                tokens: tokens,
                toolCalls: toolCalls,
                tag: "mistral.chat.complete:\(model)"))
            return resp
        }
    }
}

File: IAIso-v5.0/core/iaiso-rust/crates/middleware/src/mistral.rs
//! Mistral API wrapper.

use crate::MiddlewareError;
use iaiso_core::{BoundedExecution, StepInput, StepOutcome};
use serde::Deserialize;
use serde_json::Value;

#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Usage {
    #[serde(default)]
    pub prompt_tokens: u64,
    #[serde(default)]
    pub completion_tokens: u64,
    #[serde(default)]
    pub total_tokens: u64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ToolCall {
    pub id: String,
}

#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Message {
    #[serde(default)]
    pub tool_calls: Vec<ToolCall>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Choice {
    #[serde(default)]
    pub message: Message,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Response {
    #[serde(default)]
    pub model: String,
    #[serde(default)]
    pub usage: Usage,
    #[serde(default)]
    pub choices: Vec<Choice>,
}

pub trait Client: Send + Sync {
    fn chat_complete(&self, params: &Value) -> Result<Response, String>;
}

#[derive(Debug, Clone, Copy, Default)]
pub struct Options {
    pub raise_on_escalation: bool,
}

pub struct BoundedClient<'a, C: Client> {
    raw: C,
    execution: &'a BoundedExecution,
    opts: Options,
}

impl<'a, C: Client> BoundedClient<'a, C> {
    pub fn new(raw: C, execution: &'a BoundedExecution, opts: Options) -> Self {
        Self { raw, execution, opts }
    }

    pub fn chat_complete(&self, params: &Value) -> Result<Response, MiddlewareError> {
        match self.execution.check() {
            StepOutcome::Locked => return Err(MiddlewareError::Locked),
            StepOutcome::Escalated if self.opts.raise_on_escalation => {
                return Err(MiddlewareError::EscalationRaised)
            }
            _ => {}
        }
        let resp = self
            .raw
            .chat_complete(params)
            .map_err(MiddlewareError::Provider)?;
        let mut tokens = resp.usage.total_tokens;
        if tokens == 0 {
            tokens = resp.usage.prompt_tokens + resp.usage.completion_tokens;
        }
        let tool_calls: u64 = resp
            .choices
            .iter()
            .map(|c| c.message.tool_calls.len() as u64)
            .sum();
        let model = if resp.model.is_empty() {
            "unknown".to_string()
        } else {
            resp.model.clone()
        };
        let _ = self.execution.record_step(StepInput {
            tokens,
            tool_calls,
            tag: Some(format!("mistral.chat.complete:{}", model)),
            ..Default::default()
        });
        Ok(resp)
    }
}
